building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+
+Can you give some reasons why CMake is great?
+
+> Of course:
+
+- CMake can generate many build systems: make, ninja, VSCode project, Eclipse project, ...
+- Many GUIs and TUIs: ccmake, cmake-gui, integration in probably nearly all IDEs, ...
+- CMake is *"cross-platform"*: you can ideally use same cmake file in all OS's (easy to distinguish platform-specific things)
+- Build directory becomes independent of source directory (build multiple versions with different dependencies etc.)
+- CMake respects choices in user environment (e.g. user defines cpp compiler through CXX)
+
+### CMake "Hello World"
+
+Example adapted from [Yanson Tech](https://www.youtube.com/watch?v=wl2Srog-j7I)
+Go back to original `main.cpp`
+
+`CMakeLists.txt`:
+
+```cmake
+cmake_minimum_required(VERSION "3.12")
+
+project("HelloWorld")
+
+add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- CMake variable values: `${}`, use quotations to avoid problems with spaces
+- `mkdir build && cd build && cmake ..` ... explain files (TODO), look at makefile
+    - `CMakeCache.txt`
+- `make` and `./HelloWorld`
+
+### Multiple files
+
+- Add again `sse.cpp` and `sse.hpp` in subfolder `sse` and adapt `main.cpp`
+
+`CMakeLists.txt`:
+
+```diff
++ add_executable("${PROJECT_NAME}" main.cpp sse/sse.cpp)
+- add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- Generate and build again, this time both with verbose output TODO
+
+- What if there are multiple source files in `sse` folder?
+
+`CMakeLists.txt`:
+
+```diff
++ file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS main.cpp sse/*.cpp)
++ add_executable("${PROJECT_NAME}" "${SRC_FILES}")
+- add_executable("${PROJECT_NAME}" main.cpp sse.cpp)
+```
+
+- `GLOB_RECURSE`: filenames with wildcard characters, recurse -> also subdirectories
+- Actually discouraged since wildcard populated while generating. You need to run `cmake` again when you add a file. See [cmake docs](https://cmake.org/cmake/help/latest/command/file.html#filesystem).
+- `CONFIGURE_DEPENDS` partial remedy, but costly and no guarantee that all build systems support this
+
+### CTest
+
+- Let's add some dummy tests
+
+`main.cpp`:
+
+```diff
++  if (argc > 1)
++  {
++    return 1;
++  }
+```
+
+`CMakeLists.txt`:
+
+```diff
++ include(CTest)
++
++ add_test("test1" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}")
++ add_test("test2" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}" "someparam")
+```
+
+- `include` includes a module, some module are builtin in CMake
+- Creates target `test`
+- `make test`, explain output
+- We will learn later (after Christmas) how to work with proper packages for testing (e.g. boost)
+
+### CPack
+
+- We can also package code with cmake (or better: steer from cmake)
+
+`CMakeLists.txt`:
+
+```diff
+include(CPack)
+
+install(TARGETS "${PROJECT_NAME}" DESTINATION bin)
+```
+
+- Goes to `$CMAKE_INSTALL_PREFIX/bin`
+- Run cmake and `make package`, creates `HelloWorld-0.1.1-Linux.tar.gz`
+- We will see later (next week) how to package a `.deb`
```

Comment:
This will be difficult to get right.

Reply:
In what way will this be difficult to get right? Could you explain this a bit more in detail?

Reply:
CMake does not provide a complete solution to write Debian packages.
Creating correct packages is complex and requires a lot of additional work. (Try running `linitan` on the results. )
The CMake documentation is also be a bit cryptic on what exact information has to be provided. 
Maybe the time can be used more wisely on something else.

If correctness doesn't matter, then neither does my comment.